$OpenBSD: patch-g10_call-agent_c,v 1.1 2010/11/12 11:58:36 dcoppa Exp $

Patch from upstream: allow more hash algorithms when using scdaemon.
Needed to use SHA2-family of functions with OpenPGPv2 cards, that do
support them.

--- g10/call-agent.c.orig	Wed Feb 17 09:55:45 2010
+++ g10/call-agent.c	Wed Nov  3 14:59:39 2010
@@ -892,6 +892,23 @@ membuf_data_cb (void *opaque, const void *buffer, size
   return 0;
 }
   
+
+static const char *
+hash_algo_option (int algo)
+{
+  switch (algo)
+    {
+    case GCRY_MD_RMD160: return "--hash=rmd160 ";
+    case GCRY_MD_SHA1  : return "--hash=sha1 ";
+    case GCRY_MD_SHA224: return "--hash=sha224 ";
+    case GCRY_MD_SHA256: return "--hash=sha256 ";
+    case GCRY_MD_SHA384: return "--hash=sha384 ";
+    case GCRY_MD_SHA512: return "--hash=sha512 ";
+    case GCRY_MD_MD5   : return "--hash=md5 ";
+    default:             return "";
+    }
+}
+
 /* Send a sign command to the scdaemon via gpg-agent's pass thru
    mechanism. */
 int
@@ -938,8 +955,7 @@ agent_scd_pksign (const char *serialno, int hashalgo,
   else
 #endif
     snprintf (line, DIM(line)-1, "SCD PKSIGN %s%s",
-              hashalgo == GCRY_MD_RMD160? "--hash=rmd160 " : "",
-              serialno);
+              hash_algo_option (hashalgo), serialno);
   line[DIM(line)-1] = 0;
   rc = assuan_transact (agent_ctx, line, membuf_data_cb, &data,
                         default_inq_cb, NULL, NULL, NULL);
